### PR TITLE
Tf/categorical prediction v2

### LINKED ISSUE #393: Improve the design of negative-sampling
We want to improve the user experience of applying negative sampling + make it more generic to make it work for ranking use-cases as well.

In [Cross-Batch Negative Sampling for Training Two-Tower Recommenders](https://arxiv.org/abs/2110.15154) they describe a strategy where they start with in-batch negative sampling and then add cross-batch after _n_ steps. This is tricky to do in our current API since negative-sampling is part of the model. 

We are proposing a new type of block to make this easier: `PreLossBlock`. Since this is only used in the training/evaluation procedure (just like losses & metrics) we propose to make this part of `model.compile`. The API would then become as follows:

**Retrieval case:**
```python
model = TwoTowerModel(schema, MLPBlock([512, 128]))
model.compile(pre_loss=ContrastiveLearning(schema, "in-batch"))
model.save(save_pre_loss=False)   # user can choose whether or not to save the pre_loss

# Model can easily be converted to a youtube-dnn style retrieval model
# The model then only takes in user-features and holds all item-embeddings in an embedding-table
new_model = model.to_item_recommender(train)
model.compile(pre_loss=ContrastiveLearning(schema, "in-batch", "cross-batch"))
```

**Ranking case:**
```python
pred_task = NextItemPredictionTask(schema)
model = DLRMModel(schema, 128, MLPBlock([512, 128]), prediction_tasks=pred_task)
model.compile(pre_loss=ContrastiveLearning(schema, "popularity-based"))
```

### PR DESCRIPTION
**Pull request by sararb**
Fixes #393

### Goals :soccer:

- Refactor the prediction block related to multi-classification and next item prediction task. 
- The new prediction block also supports contrastive learning for sampled softmax support.  

### Implementation Details :construction:

- [x] Add CategoricalPrediction layer : This new class allows the definition of multi-classification tasks from a schema, with either a dense layer or an embedding table as an output layer (weight-tying use-case).

- [x] The class also supports negative sampling strategies (also known as sampled softmax) during training/evaluation to obtain top-k logits over a small subset of vocabulary. 

- [x] The negative sampling is handled by the new class `ContrastiveLookUps`. From a list of samples, it samples items ids and looks up their weights or embeddings. 

- [x] Weight-tying is handled via the new class EmbeddingTablePrediction. The logits are computed by multiplying the embedding weights by the hidden representations given by the neural network (`inputs`). 

- [x] Add a popularity-based sampler. It uses tf.random.log_uniform_candidate_sampler to approximate item probabilities using log_uniform (zipfisan).  The only requirement is that the item ids are decreasingly sorted by their count frequency (which is the default in the NVTabular `Categorify` op).

- [x] Update the default metrics types set in the __init__ method of the Prediction classes. We store a callable function `get_default_metrics` that returns the tuple of default metrics. We just need to serialize the `get_default_metrics` that gets called only when metrics are not set in the compile() method.  With this new implementation, the metric names are fixed and we initialize the default metrics only when needed. 

### Testing Details :mag:

- Testing new PopularityBasedSamplerV2
- Testing model with multi-classification task 
- Testing model with next item prediction task 
- Testing setting different negative sampling strategies: None, PopularitySampler,  PopularitySampler + In-batch sampling

### CODE REVIEW

File: merlin/models/tf/predictions/classification.py [hunk 1]
```diff
@@ -63,3 +95,445 @@ def __init__(
             name=name,
             **kwargs,
         )
+
+
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class CategoricalPrediction(ContrastivePredictionBlock):
```
Comment (gabrielspmoreira) [inline merlin/models/tf/predictions/classification.py:101]: @sararb I am wondering if we could make the API simpler for the most common use case, where the category to be predicted is an item id.
Do you think it would make sense to create a child class `ItemPrediction` inheriting from `CategoricalPrediction`, where `target_name` is set automatically to be the Tags.ITEM_ID, and with a boolean flag to enable or not weight_tying?

File: tests/unit/tf/predictions/test_classification.py [hunk 2]
```diff
@@ -38,3 +43,123 @@ def test_binary_prediction_block(ecommerce_data: Dataset, run_eagerly):
...
+
+    predictions = [
+        schema.select_by_name("item_id_seq"),
+        schema["item_id_seq"],
+        CategoricalTarget(schema["item_id_seq"]),
+        embeddings["item_id_seq"],
+        EmbeddingTablePrediction(embeddings["item_id_seq"]),
+    ]
+
+    for target in predictions:
```
[the remark this answers is not in the document]
Comment (sararb) [reply tests/unit/tf/predictions/test_classification.py:94]: First I wanted to use the `@pytest.mark.parametrize` decorator to run the sub-tests but the prediction layers are relying on other parameters defined within the test function: the embeddings tables and the schema. 

Is there a way to define a method taking as input the dataset and returning the tuple `schema, model, prediction layer` that we could pass as a parameter to the test?

File: tests/unit/tf/predictions/test_classification.py [hunk 3]
```diff
@@ -38,3 +43,123 @@ def test_binary_prediction_block(ecommerce_data: Dataset, run_eagerly):
...
+        sequence_combiner=tf.keras.layers.Lambda(lambda x: tf.reduce_mean(x, axis=1)),
+    )
+
+    predictions = [
+        schema.select_by_name("item_id_seq"),
+        schema["item_id_seq"],
+        CategoricalTarget(schema["item_id_seq"]),
+        embeddings["item_id_seq"],
+        EmbeddingTablePrediction(embeddings["item_id_seq"]),
+    ]
```
[the remark this answers is not in the document]
Comment (sararb) [reply tests/unit/tf/predictions/test_classification.py:92]: The main difference is whether we want to use a Keras dense layer  (`CategoricalTarget`) or the embeddings table of the feature (`EmbeddingTablePrediction` class) as the output.  But to simplify the API, users can only provide the `ColumnSchema` or the `EmbeddingTable` of the target feature and Merlin Models wrap them in the corresponding class.

File: tests/unit/tf/predictions/test_classification.py [hunk 4]
```diff
@@ -38,3 +43,123 @@ def test_binary_prediction_block(ecommerce_data: Dataset, run_eagerly):
...
+            mm.MLPBlock([32]),
+            mm.CategoricalPrediction(
+                target_layer=target,
+                target_name="item_id_seq",
+                negative_samplers=PopularityBasedSamplerV2(max_id=51996, max_num_samples=20),
+            ),
+        )
+        _, history = testing_utils.model_test(model, dataloader, run_eagerly=run_eagerly)
+
+    dataloader, schema = _next_item_loader(sequence_testing_data)
```
[the remark this answers is not in the document]
Comment (sararb) [reply tests/unit/tf/predictions/test_classification.py]: Thank you for the catch! This code was moved to `test_setting_negative_sampling_strategy` but I forgot to remove it here.

File: merlin/models/tf/predictions/classification.py [hunk 5]
```diff
@@ -63,3 +95,445 @@ def __init__(
...
+    ):
+
+        if feature is not None:
+            if isinstance(feature, Schema):
+                assert len(feature) == 1, "Schema can have max 1 feature"
+                col_schema = feature.first
+            else:
+                col_schema = feature
+
+            self.num_classes = col_schema.int_domain.max + 1
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin/models/tf/predictions/classification.py:265]: Right, shall we update it now or wait for #636 to be resolved?

File: merlin/models/tf/predictions/classification.py [hunk 6]
```diff
@@ -63,3 +95,445 @@ def __init__(
...
+            post=post,
+            logits_temperature=logits_temperature,
+            **kwargs,
+        )
+
+    def compile(self, negative_sampling=None, downscore_false_negatives=False):
+        if negative_sampling is not None:
+            if not isinstance(negative_sampling, (list, tuple)):
+                negative_sampling = [negative_sampling]
+            negative_sampling = [ItemSamplerV2.parse(s) for s in list(negative_sampling)]
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin/models/tf/predictions/classification.py]: I have just created the method `parse_negative_samplers` in sampling/base.py to remove repetitions.

File: merlin/models/tf/predictions/classification.py [hunk 7]
```diff
@@ -63,3 +95,445 @@ def __init__(
...
+class LookUpProtocol(Protocol):
+    def embedding_lookup(self, inputs, **kwargs):
+        pass
+
+    def __call__(self, *args, **kwargs):
+        pass
+
+
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class ContrastiveLookUps(Layer):
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin/models/tf/predictions/classification.py]: You're right, the method is first gathering the sampled ids and then looking for the corresponding embeddings.  I was thinking that `Contrastive` would refer to the negative sampling step and `LookUps` to the lookup one. So maybe we could rename it to `SampledLookUps`?

File: merlin/models/tf/predictions/classification.py (hunk 1 above)
Comment (sararb) [reply merlin/models/tf/predictions/classification.py:105]: The weight-tying is defined by explicitly providing the embedding table of the categorical feature (item-id in the common use-case) because the new implementation of the prediction blocks does not have access to the context layer anymore (We aim to remove the context layer from all merlin models blocks).  So having the weight tying as a boolean flag won't work as the prediction block doesn't have access to the embeddings table (during the build or the call method). 

we can maybe instead prepare an example notebook that illustrates the common use case (with item-id) ?

File: tests/unit/tf/predictions/test_classification.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (oliverholworthy) [reply tests/unit/tf/predictions/test_classification.py:93]: the parametrize fixture supports multiple values (example below). Though I don't think it helps in this scenario due to the dependencies between your variables.

````python
@pytest.mark.parametrize(
   ["arg1", "arg2", "arg3"],
   [
       [1, 2, 3],
       [4, 5, 6],
   ]
)
def test_with_multiple_params(arg1, arg2, arg3):

File: merlin/models/tf/predictions/classification.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oliverholworthy) [reply merlin/models/tf/predictions/classification.py:269]: probably stick with int_domain so that we can keep the cardinality change to one PR

File: merlin/models/tf/predictions/classification.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gabrielspmoreira) [reply merlin/models/tf/predictions/classification.py]: I like `SampledLookups`

File: merlin/models/tf/predictions/classification.py (hunk 1 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/predictions/classification.py:105]: I see @sararb . Your point makes sense. I agree we should have and example and/or doc showing how to use weight-tying

**Review by marcromeyn — APPROVED** [review]
